### PR TITLE
Data gen tutorial and output loss

### PR DESCRIPTION
**Pull request by lior-dikstein**

### CODE REVIEW

File: model_compression_toolkit/data_generation/pytorch/model_info_exctractors.py [hunk 1]
```diff
@@ -126,6 +131,27 @@ def __init__(self,
...
+        # Find the output nodes in the graph
+        for node in fx_model.graph.nodes:
+            if node.op == OUTPUT:
+                output_nodes += node.all_input_nodes
+
+        # Add hooks to the output layers if they are a Linear or Conv2d layer
+        for out in output_nodes:
+            for name, module in model.named_modules():
+                if name == out.target:
+                    if isinstance(module, Linear) or isinstance(module, Conv2d):
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/pytorch/model_info_exctractors.py:148]: Consider defining Linear/Conv2d as an input to the function. So user can add conv3d for example

File: model_compression_toolkit/data_generation/pytorch/optimization_functions/output_loss_functions.py [hunk 2]
```diff
@@ -43,13 +45,49 @@ def min_max_diff(
     return output_loss
 
 
+def regularized_min_max_diff(
+        output_imgs: Tensor,
+        activation_extractor: PytorchActivationExtractor,
+        eps: float = 1e-6) -> Tensor:
+    """
+    Calculate the minimum-maximum difference of output images.
```
Comment (lapid92) [inline model_compression_toolkit/data_generation/pytorch/optimization_functions/output_loss_functions.py:53]: Update func doc

**Review by lapid92 — APPROVED** [review]

File: model_compression_toolkit/data_generation/README.md [hunk 3]
```diff
@@ -20,37 +20,38 @@ pip install model-compression-toolkit
...
 
 ### Keras
 
 ```python
-from model_compression_toolkit.data_generation import get_tensorflow_data_generation_config, tensorflow_data_generation_experimental
-
+import model_compression_toolkit as mct
 # Set the configuration parameters for data generation
-data_gen_config = get_tensorflow_data_generation_config(n_iter=500,  # Number of iterations
+data_gen_config = mct.get_tensorflow_data_generation_config(n_iter=500,  # Number of iterations
```
Comment (elad-c) [inline model_compression_toolkit/data_generation/README.md:47]: why not move the arguments a line below, same as above?

File: model_compression_toolkit/data_generation/pytorch/constants.py [hunk 4]
```diff
@@ -12,11 +12,15 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ==============================================================================
+import torch
```
Comment (elad-c) [inline model_compression_toolkit/data_generation/pytorch/constants.py:15]: why do you import torch?

File: tutorials/notebooks/example_pytorch_data_generation.ipynb
Comment (ofirgo) [inline tutorials/notebooks/example_pytorch_data_generation.ipynb:1]: I don't know why I can't put a comment on specific lines in this file, so I summarize them here:
- In few sentences there is a missing period at the end.
- "Data**-**free" vs. "Data Free" (consistency).
- In the block for setting MCT configuration there is a comment regarding z score, but the code doesn't reference any z score related settings, so maybe remove the comment.
- Remove the hardcoded path to imagenet_validation folder (I'm guessing it was left there by mistake).

File: model_compression_toolkit/data_generation/pytorch/model_info_exctractors.py [hunk 5]
```diff
@@ -126,6 +131,27 @@ def __init__(self,
                 hook_handle = module.register_forward_hook(hook.hook)# Register the hook on the module
                 self.hook_handles.append(hook_handle)# Store the hook handle in the hook_handles list
 
+        # This code finds the output layers of the model and adds hooks to the input activation of
```
Comment (elad-c) [inline model_compression_toolkit/data_generation/pytorch/model_info_exctractors.py:134]: maybe put all this code in a function for readability?

File: model_compression_toolkit/data_generation/pytorch/constants.py (hunk 4 above)
Comment (ofirgo) [inline model_compression_toolkit/data_generation/pytorch/constants.py:15]: why?

File: model_compression_toolkit/data_generation/pytorch/optimization_functions/output_loss_functions.py [hunk 6]
```diff
@@ -43,13 +45,49 @@ def min_max_diff(
...
+    Returns:
+        Tensor: The computed minimum-maximum difference loss.
+    """
+    output_layers_inputs = activation_extractor.get_output_layer_input_activation()
+    weights_output_layers = activation_extractor.get_output_layers_weights()
+    if not isinstance(output_imgs, (list, tuple)):
+        output_imgs = [output_imgs]
+    output_loss = torch.zeros(1).to(get_working_device())
+    for output_weight, output, last_layer_input in zip(weights_output_layers, output_imgs, output_layers_inputs):
+        weights_norm = torch.linalg.norm(output_weight, dim=1)
```
Comment (elad-c) [inline model_compression_toolkit/data_generation/pytorch/optimization_functions/output_loss_functions.py:69]: Add comment that explains the calculation and it's objective\incentive

File: model_compression_toolkit/data_generation/pytorch/model_info_exctractors.py [hunk 7]
```diff
@@ -126,6 +131,27 @@ def __init__(self,
...
+        # those layers. We use torch.fx.symbolic_trace to get a static graph to find the output layers.
+        output_nodes = []
+        self.out_module_weights = []
+
+        # Find the output nodes in the graph
+        for node in fx_model.graph.nodes:
```
Comment (ofirgo) [inline model_compression_toolkit/data_generation/pytorch/model_info_exctractors.py:140]: This code can be replaced with a one-liner to create the list, consider it if you like

File: model_compression_toolkit/data_generation/pytorch/pytorch_data_generation.py [hunk 8]
```diff
@@ -142,7 +140,10 @@ def pytorch_data_generation_experimental(
         Returns:
             List[Tensor]: Finalized list containing generated images.
         """
-        # Get Data Generation functions and classes
+        # get a static graph representation of the model using torch.fx
+        fx_model = symbolic_trace(model)
+		
+		# Get Data Generation functions and classes
```
Comment (elad-c) [inline model_compression_toolkit/data_generation/pytorch/pytorch_data_generation.py:146]: align comment

**Review by elad-c — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/data_generation/pytorch/pytorch_data_generation.py (hunk 8 above)
Comment (ofirgo) [inline model_compression_toolkit/data_generation/pytorch/pytorch_data_generation.py:146]: extra tab here?

**Review by ofirgo — APPROVED** [review]
None of the requested changes is mandatory so I approved, but consider them

**Review by elad-c — APPROVED** [review]
